cmd/snapd-generator/main.c

```diff
@@ -30,30 +31,168 @@
+#define SYSTEMD_GENERATORS_SYSINIT_DIR SYSTEMD_GENERATORS_SVC_DIR "/sysinit.target.wants"
+
+static int create_early_mount(const char *normal_dir,
+			      const char *what,
+			      const char *unit_name)
 {
-	// Load /proc/1/mountinfo so that we can inspect the root filesystem.
-	sc_mountinfo *mounts SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
-	mounts = sc_parse_mountinfo("/proc/1/mountinfo");
-	if (!mounts) {
-		fprintf(stderr, "cannot open or parse /proc/1/mountinfo\n");
+	// Construct the file name for a new systemd mount unit.
+	char fname[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(fname, sizeof fname,
+			 "%s/%s.mount",  normal_dir, unit_name);
+
+	// Open the mount unit and write the contents.
+	FILE *f SC_CLEANUP(sc_cleanup_file) = NULL;
+	f = fopen(fname, "w");
+	if (!f) {
+		fprintf(stderr, "cannot write to %s: %m\n", fname);
 		return 1;
 	}
+	fprintf(f, "[Unit]\n");
+	fprintf(f, "Description=Early mount of kernel drivers tree\n");
+	fprintf(f, "DefaultDependencies=no\n");
+	fprintf(f, "After=systemd-remount-fs.service\n");
+	fprintf(f, "Before=sysinit.target\n");
+	fprintf(f, "Before=systemd-udevd.service systemd-modules-load.service\n");
+	fprintf(f, "Before=umount.target\n");
+	fprintf(f, "Conflicts=umount.target\n");
+	fprintf(f, "\n");
+	fprintf(f, "[Mount]\n");
+	fprintf(f, "What=%s\n", what);
+	fprintf(f, "Options=bind,shared\n");
+
+	// Wanted by sysinit.target.wants - create folders if needed and symlink
+
+	const char *dirs[] = {SYSTEMD_GENERATORS_SVC_DIR, SYSTEMD_GENERATORS_SYSINIT_DIR};
+	for (size_t i = 0; i < sizeof dirs/sizeof(char *); ++i) {
+		if (mkdir(dirs[i], 0755) != 0 && errno != EEXIST) {
+			fprintf(stderr, "cannot create %s directory: %m\n", dirs[i]);
+			return 1;
+		}
+	}
+
+	char target[PATH_MAX + 1] = { 0 };
+	char lnpath[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(target, sizeof target, "../%s.mount", unit_name);
+	sc_must_snprintf(lnpath, sizeof lnpath,
+			 SYSTEMD_GENERATORS_SYSINIT_DIR "/%s.mount", unit_name);
+	if (symlink(target, lnpath) != 0) {
+		fprintf(stderr, "cannot create symlink %s: %m\n", lnpath);
+		return 1;
+	}
+
+	return 0;
+}
 
-	sc_mountinfo_entry *root = find_root_mountinfo(mounts);
+static int ensure_kernel_drivers_mounts(const char *normal_dir)
+{
+	// Find active kernel name and revision by looking at what was
+	// mounted in /run/mnt/kernel by snap-bootstrap.
+
+	const char* kern_mnt_dir = "/run/mnt/kernel";
+	sc_mountinfo_entry *kern_minfo = find_dir_mountinfo(kern_mnt_dir);
+	if (!kern_minfo) {
+		// This is not UBuntu Core / hybrid, do nothing and do not fail
+		return 0;
+	}
+
+	// Mount source should be a snap
+	if (!sc_streq(kern_minfo->fs_type, "squashfs")) {
+		fprintf(stderr, "unexpected fs type (%s) for %s\n",
+			kern_minfo->fs_type, kern_mnt_dir);
+		return 1;
+	}
+
+	// We expect a loop device as source (/dev/loop<number>)
+	size_t src_len = strlen(kern_minfo->mount_source);
+	static const char dev_pref[] = "/dev/loop";
```

Comment:
Instead of matching `/dev/loop`, I would verify this is a block device and look at the major and minor. Then look into `/sys/dev/block/<major>:<minor>`. Just in case one day the loop devices are put in a different path.

Reply:
Good point, I've changed this now